src/modules/commitment/tests_impl.h

```diff
@@ -0,0 +1,230 @@
+    }
+    CHECK(secp256k1_pedersen_verify_tally(ctx, cptr, inputs, &cptr[inputs], outputs));
+    CHECK(secp256k1_pedersen_verify_tally(ctx, &cptr[inputs], outputs, cptr, inputs));
+    if (inputs > 0 && values[0] > 0) {
+        CHECK(!secp256k1_pedersen_verify_tally(ctx, cptr, inputs - 1, &cptr[inputs], outputs));
+    }
+    random_scalar_order(&s);
+    for (i = 0; i < 4; i++) {
+        secp256k1_scalar_get_b32(&blinds[i * 32], &s);
+    }
+    values[0] = INT64_MAX;
+    values[1] = 0;
+    values[2] = 1;
+    for (i = 0; i < 3; i++) {
+        CHECK(secp256k1_pedersen_commit(ctx, &commits[i], &blinds[i * 32], values[i], &secp256k1_generator_const_h, &secp256k1_generator_const_g));
+    }
+    CHECK(secp256k1_pedersen_verify_tally(ctx, &cptr[0], 1, &cptr[0], 1));
+    CHECK(secp256k1_pedersen_verify_tally(ctx, &cptr[1], 1, &cptr[1], 1));
+}
+
+#define MAX_N_GENS	30
+void test_multiple_generators(void) {
+    const size_t n_inputs = (secp256k1_rand32() % (MAX_N_GENS / 2)) + 1;
+    const size_t n_outputs = (secp256k1_rand32() % (MAX_N_GENS / 2)) + 1;
+    const size_t n_generators = n_inputs + n_outputs;
+    unsigned char *generator_blind[MAX_N_GENS];
+    unsigned char *pedersen_blind[MAX_N_GENS];
+    secp256k1_generator generator[MAX_N_GENS];
+    secp256k1_pedersen_commitment commit[MAX_N_GENS];
+    const secp256k1_pedersen_commitment *commit_ptr[MAX_N_GENS];
+    size_t i;
+    int64_t total_value;
+    uint64_t value[MAX_N_GENS];
+
+    secp256k1_scalar s;
+
+    unsigned char generator_seed[32];
+    random_scalar_order(&s);
+    secp256k1_scalar_get_b32(generator_seed, &s);
+    /* Create all the needed generators */
+    for (i = 0; i < n_generators; i++) {
+        generator_blind[i] = (unsigned char*) malloc(32);
+        pedersen_blind[i] = (unsigned char*) malloc(32);
+
+        random_scalar_order(&s);
+        secp256k1_scalar_get_b32(generator_blind[i], &s);
+        random_scalar_order(&s);
+        secp256k1_scalar_get_b32(pedersen_blind[i], &s);
+
+        CHECK(secp256k1_generator_generate_blinded(ctx, &generator[i], generator_seed, generator_blind[i]));
+
+        commit_ptr[i] = &commit[i];
+    }
+
+    /* Compute all the values -- can be positive or negative */
+    total_value = 0;
+    for (i = 0; i < n_outputs; i++) {
+        value[n_inputs + i] = secp256k1_rands64(0, INT64_MAX - total_value);
+        total_value += value[n_inputs + i];
+    }
+    for (i = 0; i < n_inputs - 1; i++) {
+        value[i] = secp256k1_rands64(0, total_value);
+        total_value -= value[i];
+    }
+    value[i] = total_value;
+
+    /* Correct for blinding factors and do the commitments */
+    CHECK(secp256k1_pedersen_blind_generator_blind_sum(ctx, value, (const unsigned char * const *) generator_blind, pedersen_blind, n_generators, n_inputs));
+    for (i = 0; i < n_generators; i++) {
+        CHECK(secp256k1_pedersen_commit(ctx, &commit[i], pedersen_blind[i], value[i], &generator[i], &secp256k1_generator_const_h));
```

Comment:
The last parameter should be `secp256k1_generator_const_g` instead of `secp256k1_generator_const_h`.

Related PR in Grin: https://github.com/mimblewimble/secp256k1-zkp/pull/44